### PR TITLE
Create Log to File and migrate sync debug functionality

### LINKED ISSUE #9144: Add uncaught exception handler to flush log to file
Add uncaught exception handler in case of a crash, to flush persistent logging if enabled

### PR DESCRIPTION
**Pull request by asoucar**
Because the Sync Debug work was done almost concurrently with the new logging setup, it makes sense to migrate immediately to avoid future tech debt. 
Functionality:

- Log to file of choice
- export file log to external URI
- Add logging and log flushing at uncaught exception
- Add file logging at some existing sync related points

Closes #9144 , #9314

### CODE REVIEW

File: core/logging/impl-file/src/commonMain/kotlin/net/thunderbird/core/logging/file/FileLogSink.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import net.thunderbird.core.logging.LogLevel
+import net.thunderbird.core.logging.LogSink
+
+interface FileLogSink : LogSink {
+    /**
+     * Exports from the logging method to the requested external file
+     *
+     * @param uriString The [String] for the URI to export the log to
+     *
+     * **/
```
Comment (wmontwe) [inline core/logging/impl-file/src/commonMain/kotlin/net/thunderbird/core/logging/file/FileLogSink.kt]: The leading `* ` could be removed

File: core/logging/impl-file/src/commonMain/kotlin/net/thunderbird/core/logging/file/FileLogSink.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    fun export(uriString: String)
+
+    /**
+     * On a crash, flushes buffer to file fo avoid log loss
+     *
+     * **/
```
Comment (wmontwe) [inline core/logging/impl-file/src/commonMain/kotlin/net/thunderbird/core/logging/file/FileLogSink.kt]: The leading `* ` could be removed

File: core/logging/impl-file/src/commonMain/kotlin/net/thunderbird/core/logging/file/FileLogSink.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    fun flushAndCloseBuffer()
+}
+
+/**
+ * A [LogSink] implementation that logs messages to a specified internal file.
+ *
+ * This sink uses the platform-specific implementations to handle logging.
+ *
+ * @param level The minimum [LogLevel] for messages to be logged.
+ * @param fileName The [String] fileName to log to
```
Comment (wmontwe) [inline core/logging/impl-file/src/commonMain/kotlin/net/thunderbird/core/logging/file/FileLogSink.kt:28]: Please add the remaining property description too.

File: legacy/core/src/main/java/com/fsck/k9/K9.kt [hunk 4]
```diff
@@ -32,7 +35,8 @@ object K9 : KoinComponent {
     private val generalSettingsManager: RealGeneralSettingsManager by inject()
     private val telemetryManager: TelemetryManager by inject()
     private val featureFlagProvider: FeatureFlagProvider by inject()
-    private val context: Context by inject()
+    private val syncDebugCompositeSink: CompositeLogSink by inject(named("syncDebug"))
+    private val syncDebugFileLogSink: FileLogSink by inject(named("syncDebug"))
```
Comment (wmontwe) [inline legacy/core/src/main/java/com/fsck/k9/K9.kt:39]: 👍

File: app-common/src/main/kotlin/net/thunderbird/app/common/BaseApplication.kt [hunk 5]
```diff
@@ -65,6 +66,7 @@ abstract class BaseApplication : Application(), WorkManagerConfiguration.Provide
         messagingListenerProvider.listeners.forEach { listener ->
             messagingController.addListener(listener)
         }
+        Thread.setDefaultUncaughtExceptionHandler(ExceptionHandler())
```
Comment (wmontwe) [inline app-common/src/main/kotlin/net/thunderbird/app/common/BaseApplication.kt]: Could you please check if this doesn't break the crash reporting with the default handler that will upload uncaught crashes to Google if Google Play Services are present on the device. I think you need to the get the current handler and pass the exception to it in your `ExceptionHandler`.

```suggestion
        val originalHandler = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler(ExceptionHandler(originalHandler))
```

and

```kotlin
class ExceptionHandler(
    private val defaultHandler: Thread.UncaughtExceptionHandler?
) : Thread.UncaughtExceptionHandler, KoinComponent {
    private val syncDebugFileLogSink: FileLogSink by inject(named("syncDebug"))

    override fun uncaughtException(t: Thread, e: Throwable) {
        Log.e("UncaughtException", e.toString(), e)
        syncDebugFileLogSink.flushAndCloseBuffer()
        // Forward to default handler
        defaultHandler?.uncaughtException(t, e)
    }
}
```

File: core/logging/impl-file/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+plugins {
+    id(ThunderbirdPlugins.Library.kmp)
```
Comment (wmontwe) [inline core/logging/impl-file/build.gradle.kts:2]: Add Mokkery here, in case you want to use mocks.

```suggestion
    id(ThunderbirdPlugins.Library.kmp)
    alias(libs.plugins.dev.mokkery)
```

File: core/logging/impl-file/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+kotlin {
+    sourceSets {
+        commonMain.dependencies {
+            implementation(projects.core.logging.api)
+        }
+        commonTest.dependencies {
+            implementation(libs.bundles.shared.jvm.test)
+        }
```
Comment (wmontwe) [inline core/logging/impl-file/build.gradle.kts]: Remove this declaration and switch to [Mokkery](https://mokkery.dev/) please, which is fully KMP ready.

File: core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+) : FileLogSink {
+
+    private val coroutineScope = CoroutineScope(coroutineContext + SupervisorJob())
+    private val writeFile = File(fileLocation, "$fileName.txt")
+    private val accumulatedLogs = ConcurrentHashMap<String, String>()
+
+    override fun log(event: LogEvent) {
+        try {
+            accumulatedLogs[convertLongToTime(event.timestamp)] = "priority = ${event.level}, ${event.message}"
+            createLogFile()
```
Comment (wmontwe) [inline core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt]: The file is written on every log event, this creates a lot of disk i/o. Maybe the writes could be done in batches or using a buffer with a flush interval. The `bufferedWriter` doesn't prevent disk i/o.

I also don't see where the `accumulatedLogs` are cleared after writing to file.

Should the file be appended to or always overwritten with the accumulated logs? The later could accumulate some memory over time.

File: core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                it.write(accumulatedLogs.entries.joinToString("\n") { it2 -> it2.key + " " + it2.value })
+            }
+        }
+        if (result.isFailure) {
+            result.exceptionOrNull()?.printStackTrace()
+        }
+    }
+
+    override fun flushAndCloseBuffer() {
+        writeFile.bufferedWriter().close()
```
Comment (wmontwe) [inline core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt]: The use of `bufferedWriter().use {` in the logsink already closes all used instances. This should rather write the `accumulatedLogs` to file ignoring any write delays, bufferings etc.

File: core/logging/impl-file/src/androidUnitTest/kotlin/AndroidFileLogSinkTest.android.kt
Comment (wmontwe) [inline core/logging/impl-file/src/androidUnitTest/kotlin/AndroidFileLogSinkTest.android.kt:1]: I think there is no need for integration tests. You could write to a temporary file and test the output when triggering logs. The `ContentResolver` and `Uri` are currently preventing tests and would require either more mocks or Robolectric. A better approach is to replace direct Android dependencies by fascades, for example a `FileSystemManager`. This would allow to write tests and help to indentify the corner cases to stabilize the current implementation.

Please also move the file to the `androidUnitTest/kotlin/net/thunderbird/core/logging/file` folder.

File: core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import net.thunderbird.core.logging.LogEvent
+import net.thunderbird.core.logging.LogLevel
+
+private const val ANDROID_LOG_TIME_FORMAT = "MM-dd-yy hh:mm:ss.SSS"
```
Comment (wmontwe) [inline core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt]: The format should follow a specification, for example ISO 8601 (yyyy-MM-dd'T'HH:mm:ss.SSS'Z'). KotlinX DateTime already has support, see below.

File: core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    }
+    private fun createLogFile() =
+        coroutineScope.launch {
+            writeToLogFile()
+        }
+    private fun convertLongToTime(long: Long): String {
+        val date = Date(long)
+        val format = SimpleDateFormat(ANDROID_LOG_TIME_FORMAT, Locale.US)
+        return format.format(date)
+    }
```
Comment (wmontwe) [inline core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt]: This should use [kotlinx-datetime](https://github.com/Kotlin/kotlinx-datetime).

```suggestion
    private fun convertLongToTime(long: Long): String {
        val instant = Instant.fromEpochMilliseconds(long)
        val dateTime = instant.toLocalDateTime(TimeZone.currentSystemDefault())
        return LocalDateTime.Formats.ISO.format(dateTime)
    }
```

**Comment by asoucar** [comment]
@wmontwe I addressed all your comments I think. I am aware there are 2 commented out tests at the moment, but I wanted to get the rest of the code up because I know the tests are failing due to coroutine testing finessing

File: core/logging/impl-file/src/jvmMain/kotlin/net/thunderbird/core/logging/file/FileLogSink.jvm.kt [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * A [LogSink] implementation that logs messages to a specified internal file.
+ *
+ * This sink uses the platform-specific implementations to handle logging.
+ *
+ * @param level The minimum [LogLevel] for messages to be logged.
+ * @param fileName The [String] fileName to log to
+ */
+
```
Comment (wmontwe) [inline core/logging/impl-file/src/jvmMain/kotlin/net/thunderbird/core/logging/file/FileLogSink.jvm.kt]: This newline could be removed

File: legacy/core/build.gradle.kts [hunk 13]
```diff
@@ -7,11 +7,14 @@ dependencies {
     api(projects.mail.common)
     api(projects.backend.api)
     api(projects.library.htmlCleaner)
+    api(libs.android.material)
```
Comment (wmontwe) [inline legacy/core/build.gradle.kts]: Is the material dependency needed?

File: core/logging/impl-file/build.gradle.kts (hunk 6 above)
Comment (wmontwe) [reply core/logging/impl-file/build.gradle.kts:2]: The test doesn't require mocks anymore, so this could be removed.

File: core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt
Comment (wmontwe) [inline core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt:1]: I think there are threading issues that could cause issues with `writeToLogFile`. You could use a Mutex to protect the sensible operations around `accumulatedLogs` and file operations.

The test could be fixed by always using `Timezone.UTC`.

I'm not a huge fan of using a coroutine job. This is hard to test and could cause race conditions. Some ideas to prevent immediate log writes as alternative to using a job:

- size based write, collect 50 log events before writing to file
- time based write, use a timestamp and a write interval to delay writing
- time and size based write 
- use a MutableSharedFlow that buffers a certain amount of logs

Comment (wmontwe) [reply core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt:1]: For a time based approach you could use `Clock` and `TestClock` to control the time during tests.

File: gradle/libs.versions.toml [hunk 14]
```diff
@@ -171,6 +173,7 @@ androidx-glance-appwidget-preview = { module = "androidx.glance:glance-appwidget
 androidx-glance-preview = { module = "androidx.glance:glance-preview", version.ref = "androidxGlance" }
 androidx-glance-material3 = { module = "androidx.glance:glance-material3", version.ref = "androidxGlanceMaterial3" }
 androidx-lifecycle-common = { module = "androidx.lifecycle:lifecycle-common", version.ref = "androidxLifecycle" }
+androidx-lifecycle-process = { module = "androidx.lifecycle:lifecycle-process" }
```
Comment (wmontwe) [inline gradle/libs.versions.toml]: Please add the `version.ref` here too, similarly to the other declarations.

File: gradle/libs.versions.toml [hunk 15]
```diff
@@ -107,6 +108,7 @@ timber = "5.0.1"
 turbine = "1.2.0"
 xmlpull = "1.0"
 zxing = "3.5.3"
+lifecycleProcess = "2.9.1"
```
Comment (wmontwe) [inline gradle/libs.versions.toml]: This could be removed as there is already a androidxLifecycle version declaration.

File: gradle/libs.versions.toml [hunk 16]
```diff
@@ -274,6 +278,7 @@ timber = { module = "com.jakewharton.timber:timber", version.ref = "timber" }
 turbine = { module = "app.cash.turbine:turbine", version.ref = "turbine" }
 xmlpull = { module = "com.github.cketti:xmlpull-extracted-from-android", version.ref = "xmlpull" }
 zxing = { module = "com.google.zxing:core", version.ref = "zxing" }
+lifecycle-process = { group = "androidx.lifecycle", name = "lifecycle-process", version.ref = "lifecycleProcess" }
```
Comment (wmontwe) [inline gradle/libs.versions.toml]: This could be removed as there is already a androidxLifecycle version declaration.

File: app-common/build.gradle.kts [hunk 17]
```diff
@@ -40,6 +41,8 @@ dependencies {
     implementation(projects.mail.protocols.imap)
 
     implementation(libs.androidx.work.runtime)
+    implementation(libs.androidx.lifecycle.process)
+    implementation(libs.lifecycle.process)
```
Comment (wmontwe) [inline app-common/build.gradle.kts]: This could be removed as there is already a androidx lifecycle dependency.

File: app-common/src/main/kotlin/net/thunderbird/app/common/feature/LoggerLifeCycleObserver.kt [hunk 18]
```diff
@@ -0,0 +1,17 @@
+package net.thunderbird.app.common.feature
+
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import net.thunderbird.core.logging.file.FileLogSink
+
+class LoggerLifeCycleObserver(val fileLogSink: FileLogSink?) : DefaultLifecycleObserver {
```
Comment (wmontwe) [inline app-common/src/main/kotlin/net/thunderbird/app/common/feature/LoggerLifeCycleObserver.kt]: Nit: `LoggerLifeCycleObserver` -> `LoggerLifecycleObserver`

File: core/logging/impl-file/src/jvmMain/kotlin/net/thunderbird/core/logging/file/JvmFileSystemManager.kt [hunk 19]
```diff
@@ -0,0 +1,13 @@
+package net.thunderbird.core.logging.file
+
+import kotlinx.io.RawSink
+
+/**
+ * Android implementation of [FileSystemManager] that uses [ContentResolver] to perform file operations.
+ */
+class JvmFileSystemManager() : FileSystemManager {
+    override fun openSink(uriString: String, mode: String): RawSink? {
+        // TODO: Implementation
```
Comment (wmontwe) [inline core/logging/impl-file/src/jvmMain/kotlin/net/thunderbird/core/logging/file/JvmFileSystemManager.kt]: Could you please add an issue to add the JVM implementation and add the link here.

File: legacy/core/build.gradle.kts [hunk 20]
```diff
@@ -12,6 +12,8 @@ dependencies {
     api(projects.core.android.account)
     api(projects.core.preference.impl)
     api(projects.core.android.logging)
+    api(projects.core.logging.implFile)
+    api(projects.core.logging.implComposite)
```
Comment (wmontwe) [inline legacy/core/build.gradle.kts:16]: Why adding the `implComposite` as `api` and `implementation` dependency?

File: core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+
+    override fun log(event: LogEvent) {
+        try {
+            coroutineScope.launch {
+                mutex.withLock {
+                    accumulatedLogs.add(
+                        "${convertLongToTime(event.timestamp)} priority = ${event.level}, ${event.message}",
+                    )
+                }
+                if (accumulatedLogs.size > 4) {
```
Comment (wmontwe) [inline core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt]: This should be a constant.

File: core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+            sink.close()
+            outputStream.close()
+        }
+    }
+
+    override fun flushAndCloseBuffer() {
+        runBlocking {
+            writeToLogFile()
+        }
+    }
```
Comment (wmontwe) [inline core/logging/impl-file/src/androidMain/kotlin/net/thunderbird/core/logging/file/AndroidFileLogSink.kt]: Hmm, I suggested `runBlocking`, but I realize now that this would be executed on the main thread, which is not ideal. I think it's better to remove it and change this to a suspend function.

Then the `LoggerLifeCycleObserver` needs to be adjusted as:

```kotlin
class LoggerLifeCycleObserver(val fileLogSink: FileLogSink?) : DefaultLifecycleObserver {
    override fun onStop(owner: LifecycleOwner) {
        super.onStop(owner)
        fileLogSink?.let {
            owner.lifecycleScope.launch {
                withContext(Dispatchers.IO) {
                    it.flushAndCloseBuffer()
                }
            }
        }
    }
}
```

and the `ExceptionHandler`:

```kotlin
class ExceptionHandler(
    private val defaultHandler: Thread.UncaughtExceptionHandler?,
) : Thread.UncaughtExceptionHandler, KoinComponent {
    private val syncDebugFileLogSink: FileLogSink by inject(named("syncDebug"))

    override fun uncaughtException(t: Thread, e: Throwable) {
        Log.e("UncaughtException", e.toString(), e)
        runBlocking {
            syncDebugFileLogSink.flushAndCloseBuffer()
        }
        defaultHandler?.uncaughtException(t, e)
    }
}
```

File: app-common/src/main/kotlin/net/thunderbird/app/common/feature/LoggerLifeCycleObserver.kt [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+    override fun onStop(owner: LifecycleOwner) {
+        super.onStop(owner)
+        fileLogSink?.flushAndCloseBuffer()
+    }
+
+    override fun onDestroy(owner: LifecycleOwner) {
```
Comment (wmontwe) [inline app-common/src/main/kotlin/net/thunderbird/app/common/feature/LoggerLifeCycleObserver.kt]: I think `onDestroy` is not needed as `onStop` is called before and for the case the app crashes there is the exception handler.

**Review by wmontwe — CHANGES_REQUESTED** [review]
Thank you, I just added some small adjustments to make it more robust.

File: legacy/core/build.gradle.kts (hunk 20 above)
Comment (asoucar) [reply legacy/core/build.gradle.kts:16]: Doing it as implementation makes the KoinModule not able to find it

**Review by wmontwe — APPROVED** [review]
Thank you!
